### PR TITLE
Fix #9873 - GetVertices is dropping vertices inconsistently, causing fatal error 

### LINKED ISSUE #9873: GetVertices is dropping vertices inconsistently, causing fatal error
Issue overview
--------------
This is another example for issue #9123, I tested the original example attached in #9123 in EnergyPlus V22.2 and it ran without fatal error. However, I have encountered a similar error with this [new example](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/TestSlimSurface.idf). 

In this example, there are two simple rectangular spaces stacked. The Ceiling/Floor between the two spaces are broken into 3 pieces by a slim surface in the middle.

image

The slim surface in the middle has 6 vertices. It is a ceiling surface for the lower space and a floor surface for the upper level.
image

When I run simulation with this example, I get the following error:
image

From the error message it looks like it dropped 2 vertices from surface "Ceiling 2" and dropped 1 vertex for surface "Floor 2" based on the EnergyPlus rule that the distance between vertices cannot be less than 0.1m, which made the two surfaces to have different number of vertices.

### Details
Some additional details for this issue (if relevant):
 - Windwos 10
 - EnergyPlus V22.2 
 - Unmethours link or helpdesk ticket number: NA

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here) [ TestSlimSurface.IDF](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/TestSlimSurface.idf)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 
- Fixes #9873
- Make popping vertex consistent, regardless of the order of the vertices

----

### develop

on develop, the algorithm consists on going through the vertex in order, checking the distance from this vertex to the next, and if < tolerance, pop the next vertex, and start iteration again.

### proposed

* We calculate the distance between the vertex and the next one (distanceThisToNext), as well as its previous one (distanceThisToPrev). 
* We find the vertex for which:
    * at least one of these two distances is less than the tolerance, and,
    * that minimizes the sum of distanceThisToPrev + distanceThisToNext

----

## Visual Explanation

On every figure, left is the starting surface, then as you move right, this is each pass of the poppedVertex iteration at https://github.com/NREL/EnergyPlus/blob/900b824d35cdd34ef500cf0eb7391408534fc0a9/src/EnergyPlus/SurfaceGeometry.cc#L9478-L9526

**For the floor, before and after, this is what happens**

ori_floor

**For the roof, in the original case:**

ori_roof

**For the roof, in the fixed case:**

after_roof

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 1]
```diff
@@ -10185,6 +10201,142 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_GetVerticesDropDuplicates)
...
+    // 2 o────────o 3          5 o────────o 4
+    //   │        │              │        │
+    //   ┤        │              │        │
+    //   │ Floor  │              │Ceiling │
+    //   ┤        │              │        │
+    //   │ 5      │              │ 2      │
+    //   ┤  o─────o 4            │  o─────o 3
+    //   │  │                    │  │
+    //   o──x──┬──┬───►          o──x─────────►
+    //  1    6          x       6    1
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:10298]: New test, from the example file given in #9873

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 2]
```diff
@@ -10072,27 +10072,25 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_GetSurfaceGroundSurfsTest)
...
+    // 4 o────────o 5          3 o────────o 2
+    //   │        │              │        │
+    //   ╵        ╵              ╵        ╵
+    //   ╵ Floor  ╵              ╵Ceiling ╵
+    //   ╵        ╵              ╵        ╵
+    // 3 │     2  │            4 │     5  │
+    //   o─────y  │              o─────y  │
+    //   │     │  │              │     │  │
+    //   └─────x──#───►          └─────x──#───►
+    //        1    6                  6    1
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:10171]: Had to adjust this existing test too. I just added a docstring and changed the compare_err_stream to match the new order in which it's popping now.

The difference between this test and the new one, is that this existing test has TWO vertex that need popping, versus one only in the new one.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 3]
```diff
@@ -9300,6 +9300,65 @@ namespace SurfaceGeometry {
...
+                itnext = std::begin(vertices);
+            }
+            const auto dist = distance(*it, *itnext);
+            distances[index++] = dist;
+            min_distance = std::min(min_distance, dist);
+            perimeter += dist;
+        }
+        if (min_distance >= tolerance) {
+            return {perimeter};
+        }
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc]: Pass one: calculate all distances between vertex N and N+1.
Return early if none will need to be popped.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 4]
```diff
@@ -9300,6 +9300,65 @@ namespace SurfaceGeometry {
...
+                poppedVertexPos = static_cast<int>(index);
+                if (distanceThisToPrev < distanceThisToNext) {
+                    keptVertexPos = prevIndex;
+                } else {
+                    keptVertexPos = static_cast<int>((index == nSides - 1) ? 0 : index + 1);
+                }
+            }
+        }
+
+        return {perimeter, poppedVertexPos, keptVertexPos};
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc]: Pass two: figure out the vertex that is coincident with its previous and/or next vertex and that minimizes the `distanceThisToNext + distanceThisToPrev`.
We also do return the keptVertexPos (which can be the previous or the next), so we can print the displayExtraWarning correctly.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 5]
```diff
@@ -9475,54 +9533,59 @@ namespace SurfaceGeometry {
...
-            while (poppedVertex) {
-                poppedVertex = false;
-                Perimeter = 0.0;
+            while (true) {
+                PopCoincidentVertexReturn const popResult = checkPopCoincidentVertex(vertices);
+                Perimeter = popResult.perimeter;
+                if (popResult.poppedVertexPos < 0) {
+                    // No pop needed, we're done
+                    break;
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc:9553]: If no pop, just grab the perimeter and break the while loop. We're done.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 6]
```diff
@@ -9475,54 +9533,59 @@ namespace SurfaceGeometry {
...
-                                                  format("Cannot Drop Vertex [{}]; Number of Surface Sides at minimum. This surface is now a "
-                                                         "degenerate surface.",
-                                                         curVertexIndex));
-                            }
-                            ++state.dataErrTracking->TotalDegenerateSurfaces;
-                            // mark degenerate surface?
-                        }
+                auto itKept = vertices.begin();
+                std::advance(itKept, popResult.keptVertexPos);
+                int const keptVertexIndex = popResult.keptVertexPos + 1;
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc:9562]: Grab the popped one, and the kept one (regardless of whether it's previous or next)

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 7]
```diff
@@ -9475,54 +9533,59 @@ namespace SurfaceGeometry {
...
+                                                                                        : (poppedVertexIndex == nSides and keptVertexIndex == 1);
+
+                    if (printPoppedFirst) {
+                        ShowContinueError(state, format("Vertex [{}]=({:.2R},{:.2R},{:.2R})", poppedVertexIndex, it->x, it->y, it->z));
+                        ShowContinueError(state, format("Vertex [{}]=({:.2R},{:.2R},{:.2R})", keptVertexIndex, itKept->x, itKept->y, itKept->z));
                     } else {
-                        Perimeter += DistanceCheck;
+                        ShowContinueError(state, format("Vertex [{}]=({:.2R},{:.2R},{:.2R})", keptVertexIndex, itKept->x, itKept->y, itKept->z));
+                        ShowContinueError(state, format("Vertex [{}]=({:.2R},{:.2R},{:.2R})", poppedVertexIndex, it->x, it->y, it->z));
+                    }
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc]: Gotta figure out which one of the two to print first. I aim to keep a consistent, natural, order, even when it "wraps around".

Given a surface with 3 vertices, this is what I expect

| poppedVertexIndex | keptVertexIndex | expectedPrintPoppedFirst |
|-------------------|-----------------|--------------------------|
| 1                 | 2               | TRUE                     |
| 2                 | 3               | TRUE                     |
| 3                 | 1               | TRUE                     |
| 1                 | 3               | FALSE                    |
| 2                 | 1               | FALSE                    |
| 3                 | 2               | FALSE                    |

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 8]
```diff
@@ -9475,54 +9533,59 @@ namespace SurfaceGeometry {
...
+                    if (state.dataGlobal->DisplayExtraWarnings) {
+                        ShowContinueError(state,
+                                          format("Cannot Drop Vertex [{}]; Number of Surface Sides at minimum. This surface is now a "
+                                                 "degenerate surface.",
+                                                 poppedVertexIndex));
                     }
+                    ++state.dataErrTracking->TotalDegenerateSurfaces;
+                    // mark degenerate surface?
+                    break;
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc]: If degenerate, we won't be able to pop now nor later, so exit.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 9]
```diff
@@ -9475,54 +9533,59 @@ namespace SurfaceGeometry {
...
+
+                if (state.dataGlobal->DisplayExtraWarnings) {
+                    ShowContinueError(state, format("Dropping Vertex [{}].", poppedVertexIndex));
                 }
+                --nSides;
+                vertices.erase(it);
+                // No need to recompute perimeter, because it'll be done in the next iteration, until no popping or degenerate happens
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc:9601]: if not degenerate, pop and do again.

**Comment by nrel-bot-2b** [comment]
@jmarrec it has been 15 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@jmarrec it has been 7 days since this pull request was last updated.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 10]
```diff
@@ -9475,54 +9532,59 @@ namespace SurfaceGeometry {
...
-                for (auto it = vertices.begin(); it != vertices.end(); ++it) {
-                    auto itnext = std::next(it);
-                    if (itnext == std::end(vertices)) {
-                        itnext = std::begin(vertices);
-                    }
+                auto it = vertices.begin();
+                std::advance(it, popResult.poppedVertexPos);
+                int const poppedVertexIndex = popResult.poppedVertexPos + 1;
 
-                    // TODO: use isAlmostEqual3Pt for consistency? (which uses 0.0127 m / 1/2inch instead of 0.01 m)
```
Comment (JasonGlazer) [inline src/EnergyPlus/SurfaceGeometry.cc:9497]: Kind of a side note but we really should use a single tolerance in  the code.

https://github.com/NREL/EnergyPlus/blob/50b8a5511ce559e5175524b943c26cc5b99d712d/src/EnergyPlus/SurfaceGeometry.cc#L13091

and other places

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 11]
```diff
@@ -9300,6 +9300,64 @@ namespace SurfaceGeometry {
...
+            Real64 &distanceThisToNext = distances[index];
+            Real64 &distanceThisToPrev = distances[prevIndex];
+            if ((distanceThisToNext >= tolerance) && (distanceThisToPrev >= tolerance)) {
+                continue;
+            }
+            Real64 const weight = distanceThisToNext + distanceThisToPrev;
+            if (weight < min_weight) {
+                min_weight = weight;
+                poppedVertexPos = static_cast<int>(index);
+                if (distanceThisToPrev < distanceThisToNext) {
```
Comment (JasonGlazer) [inline src/EnergyPlus/SurfaceGeometry.cc:9361]: If I'm following the logic correctly, this "less than" still means the problem can occur if the distances to the next and previous are exactly the same but are at different angles. I can't think of a solution to this issue that is not a bunch more complex.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 12]
```diff
@@ -9300,6 +9300,64 @@ namespace SurfaceGeometry {
...
+    struct PopCoincidentVertexReturn
+    {
+        double perimeter;
+        int poppedVertexPos = -1; // This is a STL vector position, 0-indexed
+        int keptVertexPos = -1;
+    };
+
+    PopCoincidentVertexReturn checkPopCoincidentVertex(const Array1D<Vector> &vertices)
+    {
+        constexpr double tolerance = 0.01;
```
Comment (JasonGlazer) [inline src/EnergyPlus/SurfaceGeometry.cc:9320]: See note on consistent tolerance below.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 4 above)
Comment (JasonGlazer) [reply src/EnergyPlus/SurfaceGeometry.cc]: Good approach!

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 13]
```diff
@@ -9475,54 +9532,59 @@ namespace SurfaceGeometry {
             auto &vertices = surface.Vertex;
             auto &nSides = surface.Sides;
 
-            bool poppedVertex = true;
-            while (poppedVertex) {
-                poppedVertex = false;
-                Perimeter = 0.0;
+            while (true) {
+                PopCoincidentVertexReturn const popResult = checkPopCoincidentVertex(vertices);
```
Comment (JasonGlazer) [inline src/EnergyPlus/SurfaceGeometry.cc:9548]: I hate to see infinite looping structures in the code but I think it is ok here.

**Review by JasonGlazer — COMMENTED** [review]
I think I followed the logic and it looks reasonable to me. I would suggest more comments about the overall approach since the algorithm seems unnecessarily complex now.

**Comment by nrel-bot-2c** [comment]
@Myoldmopar @jmarrec it has been 7 days since this pull request was last updated.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 10 above)
Comment (jmarrec) [reply src/EnergyPlus/SurfaceGeometry.cc:9497]: That TODO was left by me last time I was in there. So yes, I agree we probably should.

**Comment by jmarrec** [comment]
I'm always wary of adding too much comments since I've seen strong advocates against it (they fall out of date, code should be explicit, etc). Blaming would lead you here which has the walk-through and the visual plots. I'll try to add concise ones though

Added some in 3e7104012839d443a429d886a1e72b447bdffe60 @JasonGlazer  (and thanks for the review!)

**Comment by Myoldmopar** [comment]
Windows build failure is the usual unrelated issue, and otherwise CI is completely clean here.  @JasonGlazer thanks for looking it over, and @jmarrec thanks for providing the work!  I'll pull in develop and do a quick build/test but I think this is next to go in.

**Comment by Myoldmopar** [comment]
Yep all good still.  Let's get this one in now, thanks @jmarrec and @JasonGlazer

**Pull request merged**